File: Java/src/_0026_Remove_Duplicates_from_Sorted_Array/Solution.java
package _0026_Remove_Duplicates_from_Sorted_Array;

/**
 * 时间复杂度：O(n)
 * 空间复杂度：O(1)
 * 执行用时：1 ms, 在所有 Java 提交中击败了83.16%的用户
 * 内存消耗：40.2 MB, 在所有 Java 提交中击败了72.84%的用户
 */
public class Solution {

    public int removeDuplicates(int[] nums) {

        if (nums.length == 0)
            return 0;

        int j = 0;
        for (int i = 0; i < nums.length; i++)
            if (i == 0 || nums[i] != nums[i - 1])
                nums[j++] = nums[i];
        // nums[0] ~ num[j - 1] 所有a中不重复的数
        return j;
    }

    public static void main(String[] args) {

        int[] nums = {0, 0, 1, 1, 1, 2, 2, 3, 3, 4};
        System.out.println((new Solution()).removeDuplicates(nums));
    }
}

/*
    public int removeDuplicates(int[] nums) {
        if (nums.length == 0)
            return 0;

        int k = 1;  // [0...k)中的元素均为去重后的元素
        int removeValue = nums[0];
        for (int i = 1; i < nums.length; i++)
            if (nums[i] != removeValue) {
                nums[k++] = nums[i];
                removeValue = nums[i];
            }

        return k;
    }
 */

File: Python/_0026_Remove_Duplicates_from_Sorted_Array/Solution.py
# coding=utf-8
from typing import List


# 执行用时：48 ms, 在所有 Python3 提交中击败了63.98%的用户
# 内存消耗：15.8 MB, 在所有 Python3 提交中击败了64.03%的用户
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        T = 1
        k = 0
        for i in range(len(nums)):
            if k - T < 0 or nums[i] != nums[k - T]:
                nums[k] = nums[i]
                k += 1
        return k


if __name__ == "__main__":
    print(Solution().removeDuplicates([1, 1, 2]))  # 2
